### PR TITLE
feat: Add hearing date validation in hearing booking event (FPLA-2421)

### PR DESCRIPTION
**Pull request by aliveni**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/FPLA-2421

### Change description ###
Add validation to the Hearing booking event to make sure the hearing end date is after the hearing start date.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: e2e/tests/adminMangesHearings_test.js [hunk 1]
```diff
@@ -254,11 +254,11 @@ Scenario('HMCTS admin cancels and re-lists hearing', async ({I, caseViewPage, ma
 });
 
 Scenario('HMCTS admin adds past hearing', async ({I, caseViewPage, manageHearingsEventPage}) => {
-  hearingStartDate = moment().subtract(5,'m').toDate();
-  hearingEndDate = moment(hearingStartDate).subtract(5,'m').toDate();
+  hearingStartDate = moment().subtract(10,'m').toDate();
```
Comment (Jamesnelson117) [inline e2e/tests/adminMangesHearings_test.js:257]: what was wrong the orginal -5mins?

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/PastHearingDatesValidatorService.java [hunk 2]
```diff
@@ -25,6 +25,12 @@
             errors.add("Enter a valid end time");
         }
 
+        if (errors.isEmpty() && hearingStartDate != null && hearingEndDate != null
+            && !hearingEndDate.isAfter(hearingStartDate)
+        ) {
+            errors.add("The end date and time must be after the start date and time");
+        }
```
Comment (Jamesnelson117) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/PastHearingDatesValidatorService.java:32]: Not one for this ticket but we should refactor this at some point and use javax validation. Should we create a tech debt ticket?

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/PastHearingDatesValidatorService.java [hunk 3]
```diff
@@ -25,6 +25,12 @@
             errors.add("Enter a valid end time");
         }
 
+        if (errors.isEmpty() && hearingStartDate != null && hearingEndDate != null
```
Comment (Jamesnelson117) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/PastHearingDatesValidatorService.java:28]: is the errors.isEmpty() neccassary?

Comment (aliveni) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/PastHearingDatesValidatorService.java:28]: yes. we want to validate only when start and end dates are valid.

File: e2e/tests/adminMangesHearings_test.js (hunk 1 above)
Comment (aliveni) [reply e2e/tests/adminMangesHearings_test.js:257]: nothing wrong with the original `- 5 minutes`. only hearing end date needed change as it was earlier than the hearing start date and the test was failing with the new validation error.

**Review by Jamesnelson117 — DISMISSED** [review]

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/PastHearingDatesValidatorService.java (hunk 2 above)
Comment (aliveni) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/PastHearingDatesValidatorService.java:32]: Created tech debt ticket: https://tools.hmcts.net/jira/browse/FPLA-2527

**Review by nraghu32 — DISMISSED** [review]

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/PastHearingDatesValidatorServiceTest.java [hunk 4]
```diff
@@ -43,4 +47,24 @@ void shouldReturnValidationErrorsWhenHearingDateTimeIsNotValid() {
...
+    ) {
+        final List<String> validationErrors = service.validateHearingDates(hearingStartDate, hearingEndDate);
+
+        assertThat(validationErrors)
+            .containsOnly("The end date and time must be after the start date and time");
+    }
+
+    private static Stream<Arguments> validateHearingDatesSource() {
+        LocalDateTime startDateTime = LocalDateTime.of(2020, 12, 10, 10, 10, 0);
+        return Stream.of(
```
Comment (samuelzerozero) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/PastHearingDatesValidatorServiceTest.java:64]: This looks ok, but could use the closest boundary possible i.e. Testing minus/plus the minimum recoded (minutes/seconds)?

File: service/src/test/java/uk/gov/hmcts/reform/fpl/validation/validators/time/HasHearingEndDateAfterStartDateValidatorTest.java [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+import java.time.LocalDateTime;
+import java.util.List;
+import java.util.stream.Stream;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+class HasHearingEndDateAfterStartDateValidatorTest extends TimeValidatorTest {
+
+    @SpyBean
```
Comment (samuelzerozero) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/validation/validators/time/HasHearingEndDateAfterStartDateValidatorTest.java:19]: do we need a spy Bean?

File: service/src/test/java/uk/gov/hmcts/reform/fpl/validation/validators/time/HasHearingEndDateAfterStartDateValidatorTest.java [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+        List<String> errorMessages = validateGroupService.validateGroup(caseData, HearingEndDateGroup.class);
+        assertThat(errorMessages).isEmpty();
+    }
+
+    private static Stream<Arguments> invalidHearingEndDatesSource() {
+        LocalDateTime startDateTime = LocalDateTime.now();
+        return Stream.of(
+            Arguments.of("Hearing end date is before start date", startDateTime, startDateTime.minusDays(1)),
+            Arguments.of("Hearing end time is before start time", startDateTime, startDateTime.minusHours(1)),
```
Comment (samuelzerozero) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/validation/validators/time/HasHearingEndDateAfterStartDateValidatorTest.java:55]: same here // boundary test

**Review by samuelzerozero — DISMISSED** [review]

File: service/src/test/java/uk/gov/hmcts/reform/fpl/validation/validators/time/HasHearingEndDateAfterStartDateValidatorTest.java (hunk 5 above)
Comment (aliveni) [reply service/src/test/java/uk/gov/hmcts/reform/fpl/validation/validators/time/HasHearingEndDateAfterStartDateValidatorTest.java]: Spy Bean is not needed. Updated tests

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/PastHearingDatesValidatorServiceTest.java (hunk 4 above)
Comment (aliveni) [reply service/src/test/java/uk/gov/hmcts/reform/fpl/service/PastHearingDatesValidatorServiceTest.java:64]: updated test data with seconds/minutes difference

File: service/src/test/java/uk/gov/hmcts/reform/fpl/validation/validators/time/HasHearingEndDateAfterStartDateValidatorTest.java (hunk 6 above)
Comment (aliveni) [reply service/src/test/java/uk/gov/hmcts/reform/fpl/validation/validators/time/HasHearingEndDateAfterStartDateValidatorTest.java]: updated the test data

**Review by samuelzerozero — APPROVED** [review]

**Pull request merged**